lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -3397,6 +3397,24 @@ void ConformanceChecker::recordTypeWitness(AssociatedTypeDecl *assocType,
 
     auto overriddenRootConformance =
         overriddenConformance.getConcrete()->getRootNormalConformance();
+
+    // Don't record a type witness for an overridden associated type if the
+    // conformance to the corresponding inherited protocol and the current
+    // conformance have mismatching conditional requirements. This can turn out
+    // badly in two ways:
+    // - If a type witness for the overridden associated type has already been
+    //   recorded, we may hit an assertion in an attempt to overwrite it
+    //   with a different one. For example, the recorded type witness may be
+    //   invalid, whereas the other one---valid, and vice versa.
+    // - If the current conformance is more restrictive, this type witness may
+    //   not be a viable candidate for the overridden associated type.
+    auto currConformanceSig = DC->getGenericSignatureOfContext();
+    auto overriddenConformanceSig = overriddenRootConformance->getDeclContext()
```

Comment:
Can this be in a base class with a different generic signature?

Reply:
Yeah, good point. Testing several scenarios with classes left me pretty confused. For example, to my knowledge the type witness below ought to just be inherited, but instead we find `Derived.A` and crash in an attempt to overwrite the already-resolved type witness for the "overridden" associated type. Interestingly—and I wonder why, because we were recording type witnesses for overridden associated types all the same—this works with a 5.5.2 compiler.

```swift
protocol P {
  associatedtype A
}
protocol Q: P {
  associatedtype A
}

class Base {}
class Derived: Base {}

extension Base: P {
  typealias A = Bool
}
extension Derived: Q {
  typealias A = Int
}
```